### PR TITLE
[server] Add define to see what modules trigger most invalidation

### PR DESCRIPTION
**Pull request by kLabz**
Compiling (note: this also works for display requests) with `-D dump-invalidation-stats` (`X = 1`) or `-D dump-invalidation-stats=X` will generate a `dump/[target]/invalidation_stats.dump` file with some data about how many modules are being invalidated because of a given module. 

* `direct` column is the number of immediate dependencies for that module
* `total` column is the total count of modules invalidated because of that module, up to a "dependency depth" of `X`

Modules are sorted by the `total` column (`DESC`), with their dependency tree indented. Modules for which `direct` = `total` are not displayed.

Example (partial) output with `X = 10`:
```
module                                         | direct |  total |
------------------------------------------------------------------
battle.skill.SkillEval                         |     12 |    425 |
  battle.skill.ScriptedSkill                   |      1 |    407 |
    battle.Battle                              |     15 |    406 |
      st.Group                                 |     25 |    391 |
        ent.Place                              |     23 |    359 |
          st.Item                              |     79 |    334 |
            ent.Player                         |     27 |    248 |
              ent.Entity                       |     30 |    220 |
                ent.UnitView                   |      1 |    187 |
                  prefab.Npc                   |      5 |    186 |
battle.Battle                                  |     15 |    424 |
  st.Group                                     |     25 |    409 |
    ent.Place                                  |     23 |    377 |
      st.Item                                  |     79 |    352 |
        ent.Player                             |     27 |    266 |
          ent.Entity                           |     30 |    238 |
            ent.UnitView                       |      1 |    205 |
              prefab.Npc                       |      5 |    204 |
                Game                           |    179 |    197 |
                  Const                        |      3 |      4 |
                  prefab.BattleSnow            |      1 |      4 |
                  prefab.BattleDithering       |      1 |      2 |
...
```

cc @ncannasse -- not sure how well that fits your needs, can adjust

### CODE REVIEW

File: src/compiler/compilationCache.ml [hunk 1]
```diff
@@ -35,6 +35,7 @@ class context_cache (index : int) (sign : Digest.t) = object(self)
 	val files : (Path.UniqueKey.t,cached_file) Hashtbl.t = Hashtbl.create 0
 	val modules : (path,module_def) Hashtbl.t = Hashtbl.create 0
 	val binary_cache : (path,HxbData.module_cache) Hashtbl.t = Hashtbl.create 0
+	val invalidation_stats : (path,path list) Hashtbl.t = Hashtbl.create 0
```
Comment (Simn) [inline src/compiler/compilationCache.ml:38]: Is this reset somewhere?

Comment (kLabz) [reply src/compiler/compilationCache.ml:38]: Uh, no. I might not have tested everything again after moving things around...

**Comment by Simn** [comment]
This PR has conflicts, and somebody has to decide if this is actually useful or not.

**Comment by kLabz** [comment]
I will rework that at some point, this is not really useful in its current state

**Pull request closed without merging**